consensus/src/metrics_safety_rules.rs

```diff
@@ -125,3 +142,111 @@ impl TSafetyRules for MetricsSafetyRules {
+        waypoint::Waypoint,
+    };
+    use safety_rules::{ConsensusState, Error, TSafetyRules};
+
+    pub struct MockSafetyRules {
+        // number of initialize() calls
+        init_calls: i32,
+
+        // max initialize() calls to complete perform_initialize()
+        max_init_calls: i32,
+    }
+
+    impl MockSafetyRules {
+        pub fn new(init_calls: i32, max_init_calls: i32) -> Self {
+            Self {
+                init_calls,
+                max_init_calls,
+            }
+        }
+    }
+
+    impl TSafetyRules for MockSafetyRules {
+        fn consensus_state(&mut self) -> Result<ConsensusState, Error> {
+            let safety_data = SafetyData::new(0, 0, 0, 0, None);
+            let waypoint = Waypoint::new_any(&LedgerInfo::new(
+                BlockInfo::new(0, 0, HashValue::zero(), HashValue::zero(), 0, 0, None),
+                HashValue::zero(),
+            ));
+            Ok(ConsensusState::new(safety_data, waypoint, true))
+        }
+
+        fn initialize(&mut self, _: &EpochChangeProof) -> Result<(), Error> {
+            self.init_calls += 1;
+            if self.init_calls < self.max_init_calls {
+                return Err(Error::WaypointOutOfDate(
+                    (self.init_calls - 1) as u64,
+                    self.init_calls as u64,
+                    self.max_init_calls as u64,
+                    self.init_calls as u64,
+                ));
+            }
+            Ok(())
+        }
+
+        fn construct_and_sign_vote(&mut self, _: &MaybeSignedVoteProposal) -> Result<Vote, Error> {
+            unimplemented!()
+        }
+
+        fn sign_proposal(&mut self, _: &BlockData) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+
+        fn sign_timeout(&mut self, _: &Timeout) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+
+        fn sign_timeout_with_qc(
+            &mut self,
+            _: &TwoChainTimeout,
+            _: Option<&TwoChainTimeoutCertificate>,
+        ) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+
+        fn construct_and_sign_vote_two_chain(
+            &mut self,
+            _: &MaybeSignedVoteProposal,
+            _: Option<&TwoChainTimeoutCertificate>,
+        ) -> Result<Vote, Error> {
+            unimplemented!()
+        }
+
+        fn sign_commit_vote(
+            &mut self,
+            _: LedgerInfoWithSignatures,
+            _: LedgerInfo,
+        ) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+    }
+
+    #[test]
+    fn test_perform_initialize() {
+        let (_, mock_storage) = EmptyStorage::start_for_testing();
+        let mock_safety_rules =
+            MockSafetyRules::new(0 /* init_calls */, 10 /* max_init_calls */);
+        let mut metric_safety_rules =
+            MetricsSafetyRules::new(Box::new(mock_safety_rules), mock_storage);
+        metric_safety_rules.perform_initialize().unwrap();
```

Comment:
we should probably also test when it returns errors other than WaypointOutOfDate and it exits the loop

Reply:
@zekun000 Yeah, actually I did also test that case (returning errors other than WaypointOutOfDate and confirming loop exits) by manually changing the return of the mocked `initialize()`. Do you have a suggestion for how I can add another behavior of mocked `initialize()` without adding another mocked implementation of TSafetyRule?

Reply:
we can simply add a bool in the MockSafetyRules to tell it return other errors if configured

Reply:
Thanks for the suggestion @zekun000 !
Added a new test `test_perform_initialize_error` to confirm that loop terminates if `initialize()` returns any error other than `WaypointOutOfDate`.

Reply:
small nit: I prefer to use assert_ok!/assert_err! instead of unwrap/unwrap_err in tests to better convey the intentions like we did in https://github.com/diem/diem/pull/9163

Reply:
Noted. Will fix nits in the coming iteration.